google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java

```diff
@@ -0,0 +1,287 @@
+    }
+
+    /** Creates a {@code SessionId} given project, instance, database and session IDs. */
+    static SessionId of(String project, String instance, String database, String session) {
+      return new SessionId(new DatabaseId(new InstanceId(project, instance), database), session);
+    }
+
+    DatabaseId getDatabaseId() {
+      return db;
+    }
+
+    String getName() {
+      return name;
+    }
+  }
+
+  /**
+   * Encapsulates state to be passed to the {@link SpannerRpc} layer for a given session. Currently
+   * used to select the {@link io.grpc.Channel} to be used in issuing the RPCs in a Session.
+   */
+  static class SessionOption {
+    private final SpannerRpc.Option rpcOption;
+    private final Object value;
+
+    SessionOption(SpannerRpc.Option option, Object value) {
+      this.rpcOption = checkNotNull(option);
+      this.value = value;
+    }
+
+    static SessionOption channelHint(long hint) {
+      return new SessionOption(SpannerRpc.Option.CHANNEL_HINT, hint);
+    }
+
+    SpannerRpc.Option rpcOption() {
+      return rpcOption;
+    }
+
+    Object value() {
+      return value;
+    }
+  }
+
+  static Map<SpannerRpc.Option, ?> optionMap(SessionOption... options) {
+    if (options.length == 0) {
+      return Collections.emptyMap();
+    }
+    Map<SpannerRpc.Option, Object> tmp = Maps.newEnumMap(SpannerRpc.Option.class);
+    for (SessionOption option : options) {
+      Object prev = tmp.put(option.rpcOption(), option.value());
+      checkArgument(prev == null, "Duplicate option %s", option.rpcOption());
+    }
+    return ImmutableMap.copyOf(tmp);
+  }
+
+  private final class BatchCreateSessionsRunnable implements Runnable {
+    private final long channelHint;
+    private final int sessionCount;
+    private final SessionConsumer consumer;
+
+    private BatchCreateSessionsRunnable(
+        int sessionCount, long channelHint, SessionConsumer consumer) {
+      Preconditions.checkNotNull(consumer);
+      Preconditions.checkArgument(sessionCount > 0, "sessionCount must be > 0");
+      this.channelHint = channelHint;
+      this.sessionCount = sessionCount;
+      this.consumer = consumer;
+    }
+
+    @Override
+    public void run() {
+      List<SessionImpl> sessions = null;
+      int remainingSessionsToCreate = sessionCount;
+      while (remainingSessionsToCreate > 0) {
+        try {
+          sessions = internalBatchCreateSessions(remainingSessionsToCreate, channelHint);
+        } catch (Throwable t) {
+          consumer.onSessionCreateFailure(t, remainingSessionsToCreate);
+          break;
+        }
+        int numActuallyCreated = sessions.size();
+        if (numActuallyCreated == 0) {
+          // No sessions returned by the server. Give up creation to avoid an infinite loop.
+          consumer.onSessionCreateFailure(
+              newSpannerException(ErrorCode.UNKNOWN, "Server did not return any sessions"),
+              remainingSessionsToCreate);
+          break;
+        }
```

Comment:
when would this happen ?

Background context: we removed session limits and therefore this seems highly unlikely.

Reply:
This happened once while stress testing the functionality on the staging environment, so I added this as a safeguard against an infinite loop. I have not seen this happen on the production environment.

Reply:
i would still retry the request (same logic as numRequested > 0)

Exposing an unknown error is degradation/error to their subsequent read / write requests..

Reply:
Changed to retry also when the number of returned sessions is 0.